### PR TITLE
make LoggerFactory try its own ClassLoader if TCCCL fails, see #2298

### PR DESCRIPTION
**Pull request by vorburger**
attempts to use LoggerFactory.class.getClassLoader(); as a fallback if
Thread.currentThread().getContextClassLoader() fails to be able to load
the LogDelegateFactory.

This is because the use of getContextClassLoader() can be problematic in
"context ClassLoader sensitive environments", such as OSGi; see e.g.
http://njbartlett.name/2010/08/30/osgi-readiness-loading-classes.html or
http://wiki.eclipse.org/Context_Class_Loader_Enhancements

PS, minor: The renaming of delegateFactory variable to
newDelegateFactory just for clarity and to avoid IDE/javac warning re.
possible confusion with static field of same name.

Signed-off-by: Michael Vorburger <vorburger@redhat.com>

### CODE REVIEW

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 1]
```diff
@@ -46,18 +46,27 @@ public static synchronized void initialise() {
...
     if (className != null) {
-      ClassLoader loader = Thread.currentThread().getContextClassLoader();
+      ClassLoader tcccl = Thread.currentThread().getContextClassLoader();
       try {
-        Class<?> clz = loader.loadClass(className);
-        delegateFactory = (LogDelegateFactory) clz.newInstance();
-      } catch (Exception e) {
-        throw new IllegalArgumentException("Error instantiating transformer class \"" + className + "\"", e);
+        newDelegateFactory = loadClass(className, tcccl);
+      } catch (IllegalArgumentException e) {
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: Why catching an IllegalArgumentException here? ClassNotFoundException should be fine, no?

Comment (cescoffier) [reply src/main/java/io/vertx/core/logging/LoggerFactory.java]: Sorry, got it (didn't read the `loadClass` method.

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 2]
```diff
@@ -46,18 +46,27 @@ public static synchronized void initialise() {
...
-    LoggerFactory.delegateFactory = delegateFactory;
+    LoggerFactory.delegateFactory = newDelegateFactory;
+  }
+
+  private static LogDelegateFactory loadClass(String className, ClassLoader classLoader) {
+    try {
+      Class<?> clz = classLoader.loadClass(className);
+      return (LogDelegateFactory) clz.newInstance();
+    } catch (Exception e) {
+      throw new IllegalArgumentException("Error instantiating transformer class \"" + className + "\"", e);
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: why _transformer_?

**Comment by cescoffier** [comment]
I believe this PR is superseded by #2300

**Comment by vorburger** [comment]
> I believe this PR is superseded by #2300 

if you guys in Vert.x don't squash upon merge then #2300 with 2 commits including this instead of this is fine; but I wanted to keep the two things separate, and filed them separately because I'm too using to a Gerrit instead of a GitHub Pull Request workflow... I'll amend #2300 ASAP!

**Pull request closed without merging**